eng/pipelines/templates/steps/set-artifact-packages.yml

```
@@ -1,12 +1,44 @@
+  # B) explicitly present in the service directory. This repo splits the builds into two categories: "mgmt" and "dataplane". While
+  # a given directory may contain both, there will be a separate build definition to release the management packages. Due to this
+  # we need to artificially filter the packages to ensure that only the packages targeted for THAT ci.yml are built.
+  # When we merge the PR adding js - pullrequest, this code will also merge, meaning the public - <service> - ci builds will still operate
+  # EXACTLY the same as they did before the pipelinev3 change. This is important to ensure that we don't accidentally build the wrong packages
+  # while in the integration period. After we disable all the public `js - <service> - ci` builds, only the `js - pullrequest` build WONT provide
+  # an artifact list, which will allow the expand/contract. Meanwhile the internal builds will continue to provide the artifact list from their
+  # individual ci.yml and ci.mgmt.yml files.
   - pwsh: |
       $artifacts = '${{ convertToJson(parameters.Artifacts) }}' | ConvertFrom-Json
-      $packages = ""
-      foreach ($artifact in $artifacts)
-      {
-        $packages += "$($artifact.name),"
+      $knownArtifacts = @()
+      if ($artifacts) {
+        $knownArtifacts = $artifacts | ForEach-Object { $knownArtifacts += $_.name }
       }
-      echo "##vso[task.setvariable variable=ArtifactPackageNames]$packages"
-    displayName: "Find Packages to build"
+
+      $packageProperties = Get-ChildItem -Recurse "${{ parameters.PackageInfo }}" *.json `
+        | Where-Object { if($knownArtifacts) { $knownArtifacts -contains $_.Name.Replace(".json", "") } else { $true } }
+
+      $pkgNames = $packageProperties | Foreach-Object { $_.Name } `
+        | ForEach-Object { $_.Replace(".json", "") }
+
+      $changedServicesArray = $packageProperties | Foreach-Object { Get-Content -Raw -Path $_.FullName | ConvertFrom-Json } `
+        | ForEach-Object { $_.ServiceDirectory } | Get-Unique
+
+      $setting = $pkgNames -join ","
+      $changedServices = $changedServicesArray -join " "
+      $commaChangedServices = $changedServicesArray -join ","
+      Write-Host "Setting ChangedServices to : `n$changedServices"
+      Write-Host "##vso[task.setvariable variable=ChangedServices;]$changedServices"
+      Write-Host "##vso[task.setvariable variable=CChangedServices;]$commaChangedServices"
+      Write-Host "Setting ArtifactPackageNames to: `n$setting"
+      Write-Host "##vso[task.setvariable variable=ArtifactPackageNames;]$setting"
+    displayName: Resolve Targeted Packages
+    condition: eq(variables['ArtifactPackageNames'], '')
+
+  - pwsh: |
+      Write-Host "This run is targeting: `n$(ArtifactPackageNames) in [$(ChangedServices)]"
+    displayName: Resolve Targeted Packages
+    condition: ne(variables['ArtifactPackageNames'], '')
```

```suggestion
    condition: and(ne(variables['ChangedServices'], ''), ne(variables['ArtifactPackageNames'], ''))
```

Dunno if this is redundant or not, but in the name of preventing more powershell syntax errors when vars aren't set...

---

This plays into how I'm going to be resolving `ChangedServices` when the batch has been set for the matrix. I won't take this suggestion due to commits that have been added since you reviewed.

---

Sounds good